lightningd/plugin_hook.c

```diff
@@ -434,6 +434,21 @@ static struct hook_node *find_hook(struct hook_node *graph, const char *name)
 	return NULL;
 }
 
+/* We maintain load order for ready-to-be-added hooks. */
+static void add_to_s(struct hook_node ***s, struct hook_node *n)
+{
+	size_t pos, s_num;
+
+	s_num = tal_count(*s);
+	for (pos = 0; pos < s_num; pos++) {
+		if ((*s)[pos]->hook->plugin->index > n->hook->plugin->index)
+			break;
+	}
+	tal_resize(s, s_num + 1);
+	memmove(*s + pos+1, *s + pos, sizeof(**s) * (s_num - pos));
+	(*s)[pos] = n;
```

Comment:
I don't quite follow why this is working: we're striking dependencies off of the graph node until we finally find that it has no more, then add it to the list. However, before this change we added it to the end of the list, which was always safe (no dependency could jump over us). After this change we suddenly use insertion sort on the `index` at a potentially earlier position in the list, meaning we could end up being inserted before one of our dependencies. What am I missing here?

Reply:
You're right!

We should instead gather all the things we are going to add from one edge, and order *those* then add as a group.  That means if B and C depend on A, we append A, then sort (B,C) and append those.

I will write a test which triggers this first...